File: node/ts/test/chat/UnauthProfilesServiceTest.ts
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { config, expect, use } from 'chai';
import chaiAsPromised from 'chai-as-promised';

import * as Native from '../../Native.js';
import * as util from '../util.js';
import { TokioAsyncContext, UnauthProfilesService } from '../../net.js';
import { connectUnauth } from './ServiceTestUtils.js';
import { Aci, Pni } from '../../Address.js';

use(chaiAsPromised);

util.initLogger();
config.truncateThreshold = 0;

describe('UnauthProfilesService', () => {
  describe('accountExists', () => {
    it('faithfully returns true or false', async () => {
      const tokio = new TokioAsyncContext(Native.TokioAsyncContext_new());
      const [chat, fakeRemote] = connectUnauth<UnauthProfilesService>(tokio);

      const aci = Aci.fromUuid('9d0652a3-dcc3-4d11-975f-74d61598733f');
      const pni = Pni.fromUuid('796abedb-ca4e-4f18-8803-1fde5b921f9f');

      for (const testCase of [
        { serviceId: aci, found: true },
        { serviceId: pni, found: true },
        { serviceId: aci, found: false },
        { serviceId: pni, found: false },
      ]) {
        const responseFuture = chat.accountExists({
          account: testCase.serviceId,
        });
        const request = await fakeRemote.assertReceiveIncomingRequest();
        expect(request.verb).to.eq('HEAD');
        expect(request.path).to.eq(
          `/v1/accounts/account/${testCase.serviceId.getServiceIdString()}`
        );
        fakeRemote.sendReplyTo(request, {
          status: testCase.found ? 200 : 404,
          message: testCase.found ? 'OK' : 'Not Found',
        });
        expect(await responseFuture).to.eq(testCase.found);
      }
    });
  });
});

File: java/client/src/test/java/org/signal/libsignal/net/UnauthProfilesServiceTest.kt
//
// Copyright 2026 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.net

import org.junit.Assert.assertEquals
import org.junit.Test
import org.signal.libsignal.internal.TokioAsyncContext
import org.signal.libsignal.protocol.ServiceId
import java.util.UUID
import java.util.concurrent.TimeUnit
import kotlin.arrayOf
import kotlin.test.assertIs

class UnauthProfilesServiceTest {
  @Test
  fun testAccountExists() {
    data class TestCase(
      val serviceId: ServiceId,
      val found: Boolean,
    )

    val aci = ServiceId.Aci(UUID.fromString("9d0652a3-dcc3-4d11-975f-74d61598733f"))
    val pni = ServiceId.Pni(UUID.fromString("796abedb-ca4e-4f18-8803-1fde5b921f9f"))

    val tokioAsyncContext = TokioAsyncContext()
    val (chat, fakeRemote) =
      UnauthenticatedChatConnection.fakeConnect(
        tokioAsyncContext,
        NoOpListener(),
        Network.Environment.STAGING,
      )

    val profilesService = UnauthProfilesService(chat)

    for (testCase in listOf(
      TestCase(aci, true),
      TestCase(pni, true),
      TestCase(aci, false),
      TestCase(pni, false),
    )) {
      val responseFuture = profilesService.accountExists(testCase.serviceId)
      val (request, requestId) = fakeRemote.getNextIncomingRequest().get(1, TimeUnit.SECONDS)
      assertEquals("HEAD", request.method)
      assertEquals("/v1/accounts/account/${testCase.serviceId.toServiceIdString()}", request.pathAndQuery)
      fakeRemote.sendResponse(
        requestId,
        if (testCase.found) 200 else 404,
        if (testCase.found) "OK" else "Not Found",
        arrayOf(),
        ByteArray(0),
      )
      val result = responseFuture.get()
      val successResult = assertIs<RequestResult.Success<Boolean>>(result)
      assertEquals(testCase.found, successResult.result)
    }
  }
}
